services/openvpn/service/manager.go

```diff
@@ -264,50 +266,77 @@ func (m *Manager) ProvideConfig(_ string, sessionConfig json.RawMessage) (*sessi
-		err := json.Unmarshal(sessionConfig, &consumerConfig)
-		if err != nil {
-			return nil, errors.Wrap(err, "could not parse consumer config")
-		}
-
-		if len(consumerConfig.Ports) == 0 {
-			cp, err := m.natPingerPorts.Acquire()
+	// TODO this backward compatibility block needs to be removed once we will fully migrate to the p2p communication.
+	if conn == nil {
+		// Older clients do not send any sessionConfig, but we should keep back compatibility and not fail in this case.
+		if sessionConfig != nil && len(sessionConfig) > 0 && m.natPinger.Valid() {
+			var consumerConfig openvpn_service.ConsumerConfig
+			err := json.Unmarshal(sessionConfig, &consumerConfig)
 			if err != nil {
-				return nil, err
+				return nil, errors.Wrap(err, "could not parse consumer config")
 			}
 
-			consumerConfig.Ports = []int{cp.Num(), cp.Num(), cp.Num(), cp.Num()}
-			vpnConfig.LocalPort = cp.Num()
-		}
-
-		if m.behindNAT(publicIP) && m.portMappingFailed() {
-			for range consumerConfig.Ports {
-				pp, err := m.natPingerPorts.Acquire()
+			if len(consumerConfig.Ports) == 0 {
+				cp, err := m.natPingerPorts.Acquire()
 				if err != nil {
 					return nil, err
 				}
 
-				vpnConfig.Ports = append(vpnConfig.Ports, pp.Num())
-				vpnConfig.RemotePort = pp.Num()
+				consumerConfig.Ports = []int{cp.Num(), cp.Num(), cp.Num(), cp.Num()}
+				vpnConfig.LocalPort = cp.Num()
 			}
 
-			// For OpenVPN only one running NAT proxy required.
-			if consumerConfig.IP == "" {
-				return nil, errors.New("remote party does not support NAT Hole punching, public IP is missing")
-			}
+			if m.behindNAT(publicIP) && m.portMappingFailed() {
+				for range consumerConfig.Ports {
+					pp, err := m.natPingerPorts.Acquire()
+					if err != nil {
+						return nil, err
+					}
 
-			traversalParams.IP = consumerConfig.IP
-			traversalParams.LocalPorts = vpnConfig.Ports
-			traversalParams.RemotePorts = consumerConfig.Ports
-			traversalParams.ProxyPortMappingKey = openvpn_service.ServiceType
+					vpnConfig.Ports = append(vpnConfig.Ports, pp.Num())
+					vpnConfig.RemotePort = pp.Num()
+				}
+
+				// For OpenVPN only one running NAT proxy required.
+				if consumerConfig.IP == "" {
+					return nil, errors.New("remote party does not support NAT Hole punching, public IP is missing")
+				}
+
+				traversalParams.IP = consumerConfig.IP
+				traversalParams.LocalPorts = vpnConfig.Ports
+				traversalParams.RemotePorts = consumerConfig.Ports
+				traversalParams.ProxyPortMappingKey = openvpn_service.ServiceType
+			}
+		}
+	} else {
+		localPort, err := net.ResolveUDPAddr("udp4", fmt.Sprintf("127.0.0.1:%d", m.vpnServerPort))
+		if err != nil {
+			return nil, err
 		}
-	}
 
+		openVPNProxy, _ := net.DialUDP("udp", nil, localPort)
+		go copyStreams(openVPNProxy, conn)
+		go copyStreams(conn, openVPNProxy)
+	}
 	return &session.ConfigParams{SessionServiceConfig: vpnConfig, TraversalParams: traversalParams}, nil
 }
 
+func copyStreams(dstConn *net.UDPConn, srcConn *net.UDPConn) {
```

Comment:
Can we move this generic copy function outside openvpn service manager? I understand that it used by openvpn only atm.

Reply:
I moved this to the new file. You are right it's used only for OpenVPN transport now. I'd want to have it close to the place where it's used until we will have another need for it.